src/components/application_manager/rpc_plugins/vehicle_info_plugin/src/commands/mobile/subscribe_vehicle_data_request.cc

```diff
@@ -0,0 +1,506 @@
+        app->SubscribeToIVI(*key);
+      }
+    }
+  }
+
+  UnsubscribeFailedSubscriptions(app, message[strings::msg_params]);
+
+  if (!vi_already_subscribed_by_another_apps_.empty() ||
+      !vi_already_subscribed_by_this_app_.empty()) {
+    AddAlreadySubscribedVI(
+        const_cast<smart_objects::SmartObject&>(message[strings::msg_params]));
+  }
+
+  SendResponse(is_succeeded,
+               result_code,
+               response_info.empty() ? NULL : response_info.c_str(),
+               &(message[strings::msg_params]));
+#endif  // #ifdef HMI_DBUS_API
+}
+
+bool SubscribeVehicleDataRequest::Init() {
+  hash_update_mode_ = HashUpdateMode::kDoHashUpdate;
+  return true;
+}
+
+void SubscribeVehicleDataRequest::AddAlreadySubscribedVI(
+    smart_objects::SmartObject& msg_params) const {
+  LOG4CXX_AUTO_TRACE(logger_);
+  using namespace mobile_apis;
+  auto vi_to_string =
+      [](const mobile_apis::VehicleDataType::eType vehicle_data) {
+        for (auto& vi_str_to_int_pair : MessageHelper::vehicle_data()) {
+          if (vehicle_data == vi_str_to_int_pair.second) {
+            return vi_str_to_int_pair.first;
+          }
+        }
+        return std::string();
+      };
+
+  app_mngr::VehicleInfoSubscriptions::const_iterator it_same_app =
+      vi_already_subscribed_by_this_app_.begin();
+  for (; vi_already_subscribed_by_this_app_.end() != it_same_app;
+       ++it_same_app) {
+    msg_params[vi_to_string(*it_same_app)][strings::result_code] =
+        VehicleDataResultCode::VDRC_DATA_ALREADY_SUBSCRIBED;
+    msg_params[vi_to_string(*it_same_app)][strings::data_type] = *it_same_app;
+  }
+
+  app_mngr::VehicleInfoSubscriptions::const_iterator it_another_app =
+      vi_already_subscribed_by_another_apps_.begin();
+  for (; vi_already_subscribed_by_another_apps_.end() != it_another_app;
+       ++it_another_app) {
+    msg_params[vi_to_string(*it_another_app)][strings::result_code] =
+        VehicleDataResultCode::VDRC_SUCCESS;
+    msg_params[vi_to_string(*it_another_app)][strings::data_type] =
+        *it_another_app;
+  }
+}
+
+void SubscribeVehicleDataRequest::UnsubscribeFailedSubscriptions(
+    ApplicationSharedPtr app,
+    const smart_objects::SmartObject& msg_params) const {
+  LOG4CXX_AUTO_TRACE(logger_);
+  const VehicleData& vehicle_data = MessageHelper::vehicle_data();
+  VehicleData::const_iterator it = vehicle_data.begin();
+
+  for (; vehicle_data.end() != it; ++it) {
+    if (msg_params.keyExists(it->first)) {
+      if (msg_params[it->first][strings::result_code].asInt() !=
+          hmi_apis::Common_VehicleDataResultCode::VDRC_SUCCESS) {
+        LOG4CXX_DEBUG(logger_,
+                      "Subscription for VehicleDataType "
+                          << it->first
+                          << " is unsuccessfull. "
+                             "Unsubscribing app with connection key "
+                          << connection_key() << " from it.");
+        app->UnsubscribeFromIVI(it->second);
```

Comment:
UnsubscribeFromIVI shall be a function in the new plugin

Reply:
@yang1070 b5a91fdd09dfd4d1620cce0bd63979ceb299703a